src/background.js

```diff
@@ -9,60 +9,81 @@
 import moment from 'moment';
 import { times, timeForId } from './lib/times';
 
-browser.runtime.onMessage.addListener(({op, message}) => {
-  console.log('backgroundMessage', op, message);
-  switch(op) {
-    case 'schedule':
-      return handleSchedule(message);
-    case 'cancel':
-      return handleCancel(message);
-    case 'update':
-      return handleUpdate(message);
-  }
-});
+const DEBUG = (process.env.NODE_ENV === "development");
+const WAKE_ALARM_NAME = 'snooze-wake-alarm';
 
-function handleSchedule(message) {
-  let item = {};
-  item[`${message.time}`] = message;
-  browser.storage.local.set(item);
-  browser.alarms.create(`${message.time}`, {when: message.time});
+function log(...args) {
+  if (DEBUG) { console.log('SnoozeTabs (BE):', ...args); }
 }
 
-function handleCancel(message) {
-  const id = `${message.time}`;
-  browser.storage.local.remove(id);
-  browser.alarms.clear(id);
+function init() {
+  log('init()');
+  browser.alarms.onAlarm.addListener(handleWake);
+  browser.notifications.onClicked.addListener(handleNotificationClick);
+  browser.runtime.onMessage.addListener(handleMessage);
+  handleWake().then(updateWakeAlarm);
 }
 
-function handleUpdate(message) {
-  handleCancel(message.old);
-  handleSchedule(message.updated);
+function handleMessage({op, message}) {
+  log('backend received', op, message);
+  if (messageOps[op]) { messageOps[op](message); }
 }
 
-browser.notifications.onClicked.addListener(notificationId => {
-  let [windowId, tabId] = notificationId.split(':');
-  browser.windows.update(+windowId, {focused: true});
-  browser.tabs.update(+tabId, {active: true});
-});
-
-browser.alarms.onAlarm.addListener(alarm => {
-  browser.storage.local.get(alarm.name).then(messages => {
-    let message = messages[alarm.name]
-    browser.tabs.create({
-      active: false,
-      url: message.url,
-      windowId: message.windowId
-    }).then(tab => {
-      browser.notifications.create(`${message.windowId}:${tab.id}`, {
+const idForItem = item => `${item.time}`;
+
+const messageOps = {
+  schedule: message => {
+    const toSave = {};
+    toSave[idForItem(message)] = message;
+    return browser.storage.local.set(toSave).then(updateWakeAlarm);
+  },
+  cancel: message =>
+    browser.storage.local.remove(idForItem(message)).then(updateWakeAlarm),
+  update: message =>
+    messageOps.cancel(message.old).then(() => messageOps.schedule(message.updated))
+};
+
+function updateWakeAlarm() {
+  return browser.alarms.clearAll()
+    .then(() => browser.storage.local.get())
+    .then(items => {
+      const times = Object.values(items).map(item => item.time);
```

Comment:
Isn't that the same as `Object.keys(items)` (because we're storing the items with their time as the key)?
(Or `items.map(k => +k)` at worst…)